File: DS using c/stack-applications/infix-postfix.c
#include <stdio.h>
#include <string.h>
#define MAX_SIZE 200

char stack[MAX_SIZE];
char postfix[MAX_SIZE];
int top = -1;
char pop()
{
    return stack[top--];
}
void push(char j)
{
    top++;
    stack[top] = j;
}
char peek()
{
    return stack[top];
}
int is_operand(char ch)
{
    return (ch >= 'a' && ch <= 'z') || (ch >= 'A' && ch <= 'Z') || (ch>='0' && ch<='9');
}

int precedence(char operate)
{
    if (operate == '^')
        return 3;
    else if (operate == '*' || operate == '/')
        return 2;
    else if (operate == '+' || operate == '-')
        return 1;
    return 0;
}

char *infix_to_postfix(char infix[])
{
    int k = -1;
    for (int i = 0; i < strlen(infix); i++)
    {
        if (is_operand(infix[i]))
        {
            postfix[++k] = infix[i];
        }
        else if (infix[i] == '(')
        {
            push(infix[i]);
        }
        else if (infix[i] == ')')
        {
            while (top != -1 && peek() != '(')
            {
                postfix[++k] = pop();
            }
            if (top != -1)
                pop();
        }
        else
        {
            while (top != -1 && precedence(infix[i]) <= precedence(stack[top]))
                postfix[++k] = pop();
            push(infix[i]);
        }
    }
    while (top != -1 && peek() != '(')
        postfix[++k] = pop();
    postfix[++k] = '\0';
    return postfix;
}

int main()
{
    char infix[MAX_SIZE];
    printf("Enter infix Expression: ");
    scanf("%s", infix);
    printf("Postfix expression is %s\n", infix_to_postfix(infix));
    return 0;
}

File: DS using python/stack-applications/infix-postfix.py
import array as arr


class Stack:
    def __init__(self):
        self.MAXSIZE = 100
        self.stack = arr.array("u", "\0" * self.MAXSIZE)
        self.top = -1
        self.postfix = ""

    def pop(self):
        popped = self.stack[self.top]
        self.top -= 1
        return popped

    def push(self, j):
        if self.top == self.MAXSIZE - 1:
            print("stack is full")
            exit(0)
        self.top += 1
        self.stack[self.top] = j

    def peek(self):
        return self.stack[self.top]

    def isoperand(self, ch):
        return (ch >= "a" and ch <= "z") or (ch >= "A" and ch <= "Z")

    def precedence(self, operator):
        if operator == "^":
            return 3
        elif operator == "*" or operator == "/":
            return 2
        elif operator == "+" or operator == "-":
            return 1
        return 0

    def infix_to_postfix(self, infix):
        for i in range(len(infix)):
            if self.isoperand(infix[i]) or infix[i].isdigit():
                self.postfix += infix[i]
            elif infix[i] == "(":
                self.push(infix[i])
            elif infix[i] == ")":
                while self.top != -1 and self.peek() != "(":
                    self.postfix += self.pop()
                if self.top != -1:
                    self.pop()
            else:
                while self.top != -1 and self.precedence(infix[i]) <= self.precedence(
                    self.stack[self.top]
                ):
                    self.postfix += self.pop()
                self.push(infix[i])
        while self.top != -1 and self.peek() != "(":
            self.postfix += self.pop()
        return self.postfix


if __name__ == "__main__":
    infix = input("Enter infix expression : ")
    stack = Stack()
    print("Postfix expression is", stack.infix_to_postfix(infix))
